runtime/expr/function/unflatten.go

```diff
@@ -36,49 +39,67 @@ func (u *Unflatten) Call(ctx zed.Allocator, args []zed.Value) *zed.Value {
 	u.types = u.types[:0]
 	u.values = u.values[:0]
 	for it := val.Bytes.Iter(); !it.Done(); {
-		path, typ, vb := u.parseElem(array.Type, it.Next())
+		bytes := it.Next()
+		path, typ, vb, err := u.parseElem(array.Type, bytes)
+		if err != nil {
+			return u.zctx.WrapError(err.Error(), ctx.NewValue(array.Type, bytes))
+		}
 		if typ == nil {
 			continue
 		}
-		root.addPath(&u.recordCache, path)
+		removed := root.addPath(&u.recordCache, path)
+		if removed > 0 {
+			u.types = u.types[:len(u.types)-removed]
+			u.values = u.values[:len(u.values)-removed]
+		}
 		u.types = append(u.types, typ)
 		u.values = append(u.values, vb)
 	}
 	u.builder.Reset()
 	types, values := u.types, u.values
-	typ := root.build(u.zctx, &u.builder, func() (zed.Type, zcode.Bytes) {
+	typ, err := root.build(u.zctx, &u.builder, func() (zed.Type, zcode.Bytes) {
 		typ, value := types[0], values[0]
 		types, values = types[1:], values[1:]
 		return typ, value
 	})
+	if err != nil {
+		return u.zctx.WrapError(err.Error(), &val)
+	}
 	return ctx.NewValue(typ, u.builder.Bytes())
 }
 
-func (u *Unflatten) parseElem(inner zed.Type, vb zcode.Bytes) (field.Path, zed.Type, zcode.Bytes) {
+func (u *Unflatten) parseElem(inner zed.Type, vb zcode.Bytes) (field.Path, zed.Type, zcode.Bytes, error) {
 	if union, ok := zed.TypeUnder(inner).(*zed.TypeUnion); ok {
 		inner, vb = union.SplitZNG(vb)
 	}
 	typ := zed.TypeRecordOf(inner)
 	if typ == nil || len(typ.Columns) != 2 {
-		return nil, nil, nil
+		return nil, nil, nil, nil
 	}
 	nkey, ok := typ.ColumnOfField("key")
 	if !ok {
-		return nil, nil, nil
-	}
-	if a, ok := zed.TypeUnder(typ.Columns[nkey].Type).(*zed.TypeArray); !ok && a.Type != zed.TypeString {
-		return nil, nil, nil
+		return nil, nil, nil, nil
 	}
+
 	vtyp, ok := typ.TypeOfField("value")
 	if !ok {
-		return nil, nil, nil
+		return nil, nil, nil, nil
 	}
 	it := vb.Iter()
 	kbytes, vbytes := it.Next(), it.Next()
 	if nkey == 1 {
 		kbytes, vbytes = vbytes, kbytes
 	}
-	return u.decodeKey(kbytes), vtyp, vbytes
+	ktyp := typ.Columns[nkey].Type
+	if ktyp.ID() == zed.IDString {
+		u.path = u.path[:0]
+		u.path = append(u.path, zed.DecodeString(kbytes))
+		return u.path, vtyp, vbytes, nil
+	}
+	if a, ok := zed.TypeUnder(ktyp).(*zed.TypeArray); ok && a.Type == zed.TypeString {
```

Comment:
`a.Type` might be a named type.
```suggestion
	if a, ok := zed.TypeUnder(ktyp).(*zed.TypeArray); ok && a.Type.ID() == zed.IDString {
```